### DISCUSSION TITLE
rebooting issue esp32 audio kit 2.2 / not enough memory -> memory exhausted

### DISCUSSION THREAD

**Discussion by prouting** [marcel-licence/esp32_multitrack_looper#7] [Q&A]
Hi Marcel

Sorry my ignorance in programming

I have successfully compiled by modifying
define ES8388_CFG_I2C from1 to 2

but now rebooting continuously showing this:
------------------------------
Loading data
Firmware started successfully
Connect to ES8388 codec... 0x00: 0x05
0x01: 0x40
AdcCh1!
MixChAMPL!
ES8388 setup finished!
Reg 0x00 = 0x05
Reg 0x01 = 0x40
...

Setup Serial2 with 31250 baud with rx: 21 only
Setup MidiPort2 using Serial2
Total PSRAM: 4192091
Free PSRAM: 4015179
done!..
- Free PSRAM: 2988675
done!..
- Free PSRAM: 1962171
done!..
- Free PSRAM: 935667
not enough memory!
ESP.getFreeHeap() 224596
ESP.getMinFreeHeap() 224416
ESP.getHeapSize() 255352
ESP.getMaxAllocHeap() 110580
Total heap: 255352
Free heap: 224596
Total PSRAM: 4192055
Free PSRAM: 935667
Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400d35a3  PS      : 0x00060f30  A0      : 0x800d3811  A1      : 0x3ffcb8f0  
A2      : 0x3ffc394c  A3      : 0x3ffc388c  A4      : 0x3ffc3688  A5      : 0x3ffc3ab4  
A6      : 0x3ffc3aa4  A7      : 0x3ffc3ad8  A8      : 0x00000000  A9      : 0x0000000c  
A10     : 0x3ffc3694  A11     : 0x00000003  A12     : 0x00000000  A13     : 0x00000000  
A14     : 0x00000004  A15     : 0x000000ff  SAR     : 0x0000001d  EXCCAUSE: 0x0000001d  
EXCVADDR: 0x00000000  LBEG    : 0x400d358c  LEND    : 0x400d3650  LCOUNT  : 0x00000000  

Backtrace:0x400d35a0:0x3ffcb8f00x400d380e:0x3ffcb920 0x400d385f:0x3ffcb960 0x400d6c1d:0x3ffcb980 

ELF file SHA256: 0000000000000000

Rebooting...
ets Jun  8 2016 00:22:57
--------------------------------------------

i'm doing wrong
Thanx in advance

Pedro

**Comment by marcel-licence** [comment]
Hi,

looks like there is not enough memory for one of the modules `not enough memory!`
That means when the looper wants to run it tries to work a module which couldn't allocate memory and this causes the exception and finally the reboot.

Did you modify one of the following defines?
TRACK_CNT, MAX_LOOP, MAX_DELAY

They influence the required memory. If you setup to high values the looper is not able to run

**Reply by prouting** [reply]
Hello marcel

I have managed not to get an error by lowering MAX_LOOP but there is no audio or metronome on my board, I will try other ES8388 configurations. My ESP32Audio kit board is V2.2 A149

#define MAX_LOOP    126492    /*513246*/
#define TRACK_CNT   4

Another query, teraterm doesn't come out right, it scrolls me. How to configure it?

I am also struggling with the midi, I made interface with 6n138 to IO22 and I see D4 flashing but nothing else

Thank for your patience

**Reply by marcel-licence** [reply]
Hello Pedro,

good to hear that is now running. Hope you will find the correct configuration soon. You should also get some messages indicating that the ES8388 is configured. You can also use the I2C scanner to check different PINs. The ES8388 is connected via I2C and needs to be configured to start its operation.

The teraterm window should be resizeable. I think using a bigger size should avoid scrolling (like fullscreen). Hope this helps.

D4 flashing sounds also good. You can change the MIDI mapping in z_config.ino.

Best regards,
Marcel

**Reply by prouting** [reply]
Hello

Some progress today

I think there is a bug in ML_synth_tools in board_audio_kit_es8388.h

#elif ES8388_CFG_I2C==2
#define ES8388_PIN_SDA  33
#define ES8388_PIN_SCL  32
#elif ES8388_CFG_I2C==2  <<<<<  I rename to 3
#define ES8388_PIN_SDA  27
#define ES8388_PIN_SCL  28

Compiling basic_synth for test and I get audio tone  if enable #define OUTPUT_SAW_TEST with the configuration
I2C=2   I2S=2     voilà

I can see serial midi activity if enable midi_dump 
although I don't get notes from my midi keyboard,
neither by activating #define NOTE_ON_AFTER_SETUP
:,(

**Reply by marcel-licence** [reply]
Thank you for your feedback.
I've fixed the file. 

The multitrack looper project does not include any synthesizer module. You can only activate recording and playback. To play sounds you can try one of the synthesizer projects. I used the line in with an external keyboard to record the video.

Hope this helps.

**Reply by prouting** [reply]
Hello Marcel

It already works with the analog modification keys, but the sound is very robotic and lo-fi
I will continue to investigate why

Regards

**Reply by marcel-licence** [reply]
Hello Pedro,

the robotic sound may be a result of too slow code. That might happen if you put some debug output in the main loop. It would slow down the audio processing generating audible results.

Best regards,
